### DISCUSSION TITLE
Keycloak LDAPS does not find valid certification path to requested target in Production

### DISCUSSION THREAD

**Discussion by reddybhavaniprasad** [keycloak/keycloak#17085] [Q&A]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

ldap

### Describe the bug

I have deployed **Keycloak 20** as a kubernetes pod and tried to synchronize all the users from **LDAPS**. So updated the LDAPS details in the Keycloak dashboard but the Test connection fails to LDAPS.

So tried the same with LDAP and it is successful.

There are no issues with the certificates because the same LDAPS is properly integrated in another setup without keycloak and by using the same commercial certificates.

### Version

v20.0.0

### Expected behavior

Expecting the Keycloak to accept the truststore provided and connect to LDAPS. Finally allow importing all the users from the LDAPS configured into Keycloak.

### Actual behavior

The test connection to LDAPS fails with some of the errors mentioned below,
1. Could not negotiate TLS
[keycloak_ldaps_error.log](https://github.com/keycloak/keycloak/files/10330397/keycloak_ldaps_error.log)
[keycloak_ldaps_error.log](https://github.com/keycloak/keycloak/files/10330403/keycloak_ldaps_error.log)

2. PKIX path building failed. Unable to find valid certification path to requested target. (javax.namingCommunication Exception)
3. Connection or Outbound has closed.

### How to Reproduce?

Below are the list of steps followed:
1. Created a truststore (truststore.jks) file using the root and domain certificates by running keytool command.
```
keytool -import -alias vxx.xxxxx.com -keystore /opt/keycloak/data/keystore/truststore.jks -file /opt/keycloak/data/keystore/domain.cer
keytool -import -alias root -keystore /opt/keycloak/data/keystore/truststore.jks -file /opt/keycloak/data/keystore/root.cer
```
2. Ran the kc.sh start script with the truststore (truststore.jks) file as mentioned in the official documentation,
```
bin/kc.sh start --spi-truststore-file-file=truststore.jks --spi-truststore-file-password=changeit --spi-truststore-file-hostname-verification-policy=ANY
```
**NOTE**: The above command is passed as arguments in the deployment.yaml. The truststore.jks and certificates are persisted in the persistent volume.

3. Now updated the User federation section of the Keycloak dashboard with all the details like the LDAPS endpoint, User DN, Bind DN and Bind Password.
4. After clicking the "Test Connection" button, it fails and prompts to check through the logs for test connection failure. The logs shows 2 to 3 types of errors as mentioned above.

### Anything else?

NOTE: Repeated the step-2 from above with the standard java truststore configuration,
```
keytool -import -alias vxx.xxxxx.com -keystore /etc/java/java-11-openjdk/java-11-openjdk-11.0.17.0.8-2.el8_6.x86_64/lib/security/cacerts -file /opt/keycloak/data/keystore/domain.cer
keytool -import -alias root -keystore /etc/java/java-11-openjdk/java-11-openjdk-11.0.17.0.8-2.el8_6.x86_64/lib/security/cacerts -file /opt/keycloak/data/keystore/root.cer
```
There are no issues with the certificates because the same LDAPS is properly integrated in another setup without keycloak and by using the same commercial certificates.

**Comment by hmlnarik** [comment]
Thank you for the detailed report. This is a question on how to configure LDAP truststore with quarkus. Hence I am converting this into a Q&A discussion.

**Reply by mkoppal-px** [reply]
Facing the same issue. Need help.

**Comment by konachan700** [comment]
I have the same issue.

**Comment by martinleim** [comment]
I would recommend to enable TLS debug logging by adding `-Djavax.net.debug=ALL`. One way to set this is using the environment variable `JAVA_OPTS_APPEND`.
This will print a lot of useful information, including the list of certificates in the truststore and the handshake data. 
More info about TLS debugging can be found on https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/ReadDebug.html
